indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/TwoCloudWorkerProvisioningStrategy.java

```diff
@@ -121,19 +131,30 @@ Provisioner makeProvisioner(TasksAndWorkers runner)
       @Override
       public boolean doTerminate()
       {
-        return provisioner1.doTerminate() || provisioner2.doTerminate();
+        // Always try to terminate in both clouds before returning from this method
+        boolean terminated1 = provisioner1.doTerminate();
+        boolean terminated2 = provisioner2.doTerminate();
+        return terminated1 || terminated2;
       }
 
       @Override
       public boolean doProvision()
       {
-        return provisioner1.doProvision() || provisioner2.doProvision();
+        // Always try to provision in both clouds before returning from this method
+        boolean provisioned1 = provisioner1.doProvision();
+        boolean provisioned2 = provisioner2.doProvision();
+        return provisioned1 || provisioned2;
       }
 
       @Override
       public ScalingStats getStats()
       {
-        return new ScalingStats(provisioner1.getStats(), provisioner2.getStats());
+        ScalingStats stats1 = provisioner1.getStats();
+        ScalingStats stats2 = provisioner2.getStats();
+        ScalingStats stats = new ScalingStats(stats1.size() + stats2.size());
```

Comment:
We could lose some initial events, for example, we call stats1.size() (when it is 0) than **before** we call stats1.toList() other thread can quickly add to stats1, but we already set max size 0 for a new queue, and we lose those events.